src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/UseValidPlatformString.cs

```diff
@@ -0,0 +1,241 @@
+                    !context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemRuntimeVersioningUnsupportedOSPlatformAttribute, out var unsupportedAttribute))
+                {
+                    return;
+                }
+
+                var knownPlatforms = PooledDictionary<string, int>.GetInstance(StringComparer.OrdinalIgnoreCase);
+                AddPlatformsAndVersionCountFromGuardMethods(operatingSystemType, knownPlatforms);
+                AddPlatformsFromMsBuildOptions(knownPlatforms, context.Options.GetMSBuildItemMetadataValues(
+                    MSBuildItemOptionNames.SupportedPlatform, context.Compilation, context.CancellationToken));
+
+                context.RegisterOperationAction(context => AnalyzeOperation(context.Operation, context, knownPlatforms), OperationKind.Invocation);
+                context.RegisterSymbolAction(context => AnalyzeSymbol(context.ReportDiagnostic, context.Symbol,
+                    supportedAttriubte, unsupportedAttribute, knownPlatforms, context.CancellationToken), s_symbols);
+                context.RegisterCompilationEndAction(context => AnalyzeSymbol(context.ReportDiagnostic, context.Compilation.Assembly,
+                    supportedAttriubte, unsupportedAttribute, knownPlatforms, context.CancellationToken));
+            });
+
+            static void AddPlatformsAndVersionCountFromGuardMethods(INamedTypeSymbol operatingSystemType, PooledDictionary<string, int> knownPlatforms)
+            {
+                var methods = operatingSystemType.GetMembers().OfType<IMethodSymbol>();
+                foreach (var m in methods)
+                {
+                    if (m.IsStatic &&
+                        m.ReturnType.SpecialType == SpecialType.System_Boolean &&
+                        NameAndParametersValid(m))
+                    {
+                        var versionCount = ExtractPlatformAndVersionCount(m, out var platform);
+                        if (!knownPlatforms.TryGetValue(platform, out var count) ||
+                            versionCount > count)
+                        {
+                            knownPlatforms[platform] = versionCount; // only keep highest count
+                        }
+                    }
+                }
+            }
+
+            static void AddPlatformsFromMsBuildOptions(PooledDictionary<string, int> knownPlatforms, ImmutableArray<string> msBuildPlatforms)
+            {
+                foreach (var platform in msBuildPlatforms)
+                {
+                    if (!knownPlatforms.ContainsKey(platform))
+                    {
+                        knownPlatforms.Add(platform, 4); // Default version count is 4
+                    }
+                }
+            }
+
+            static int ExtractPlatformAndVersionCount(IMethodSymbol method, out string platformName)
```

Comment:
I suggest having the return type be a `Tuple<string platformName, int versionParts>`